### PR TITLE
Automatic publishing proposal

### PR DESCRIPTION
**Pull request by omrilotan**
# NPM smart publishing

## TL;DR
| Branch type | action |
| --- | --- |
| **Feature branch** | Release RC versions on tag by branch name. |
| **Master branch** | Release clean semver on "latest" tag. Create a git tag. |

## Feature branch
- Published only if the version has a pre-release section which contains `rc`:
- Branch versions get a suffix that matches the commit ID
- Tags are named after the branch

## Master branch
- Only publishes clean semver versions, no pre-release
- Publishes versions to tag `latest`
- Creates tag in git repository

| branch | version | Will publish | to tag | Will create git tag |
| --- | --- | --- | --- | --- |
| `my_feature_branch` | `"1.3.0"` | nothing | N/A | No
| `my_feature_branch` | `"1.3.1-alpha"` | nothing | N/A | No
| `my_feature_branch` | `"1.3.1-rc"` | `1.3.1-rc-c447f6a` | `my_feature_branch` | No
| `my_feature_branch` | `"1.3.1-rc.1"` | `1.3.1-rc.1-c447f6a` | `my_feature_branch` | No
| `master` | `"1.3.0"` | `1.3.0` | `latest` | Yes
| `master` | `"1.3.0-beta"` | Throws Error | N/A | No
| `master` | `"1.3.0-rc"` | Throws Error | N/A | No

> ## Example flow:
>
> ### @fiverr/package:
> #### Developing
> `git checkout -b my_awesome_feature`
> `npm version 1.4.0-beta`
> `git push origin my_awesome_feature`
> - Circle CI works, does not release package version
>
> #### Release candidate (live testing)
> `npm version 1.4.0-rc`
> `git push origin my_awesome_feature`
> - Circle CI works, releases package version `1.4.0-rc-c447f6a` to tag `my_awesome_feature`
>
> ### Consuming Service
> #### Testing
> `npm i -S @fiverr/package@my_awesome_feature`
> - Installs `@fiverr/package` version `1.4.0-rc-c447f6a`
>
> ### @fiverr/package:
> #### Fix issue in release candidate
> `git push origin my_awesome_feature`
> - Circle CI works, releases package version `1.4.0-rc-a594081` to tag `my_awesome_feature`
>
> ### Consuming Service
> #### Testing
> `npm i -S @fiverr/package@my_awesome_feature` // same tag
> - Installs `@fiverr/package` version `1.4.0-rc-a594081` // new version
>
> ### @fiverr/package:
> #### Finalising
> `npm version 1.4.0`
> `git push origin my_awesome_feature`
> - Circle CI works, releases nothing
>
> `git checkout master`
> `git merge my_awesome_feature`
> `git push origin master`
> - Circle CI works, releases package version `1.4.0` to tag `latest`
>
> ### Consuming Service
> #### Using
> `npm i -S @fiverr/package@latest`
> - Installs `@fiverr/package` version `1.4.0`

### CODE REVIEW

**Comment by ealush** [comment]
```git push origin master```
Don't you think this should be done by CI?

**edit.**
To be fair, I believe this should be done by CI too:
`npm version 1.4.0`
You can config CI to use the pushed git tag to set the npm version.

**edit.2**

The flow I'd recommend:

**Finalising**
```
git push origin my_awesome_feature
git checkout master
git merge my_awesome_feature
git --tag -a 1.4.0 -m 1.4.0
git push origin 1.4.0
```
Then CI:
* Runs tests
* bumps version with your tag
```
npm version $GIT_TAG -m "Updating $PACKAGE_NAME: V$GIT_TAG"
npm publish
git push origin master
```

**Comment by ealush** [comment]
Regarding branching convention

My **opinion** is going to go against our notion of continuous integration, but so does following semver (to a certain extent), so I am fine with it:

I do not think that feature branches should be merged directly to master. I think it is time we adopt a more conservative approach which allows more time and room for integration tests.

I think that my assumption that is fairly accepted that frequent updates are both bad from a performance point of view (new versions loaded), and in overall system stability - as more frequent api changes mean less testing and a higher likelihood of introducing new bugs into the system.

Cumulative versioning
----

When working with packages especially, I believe it is safer to work with **cumulative versions**, rather than directly and immediately update consumers without extensive testing. We can have packages release cycle (half-monthly, for example), **when only bug fixes are pushed directly to master**.

It is not going to have any effect on development pace, as you can always use your `rc.x` tag in development while waiting for your changes to be integrated.



brancing plan
----

| Branch | Use
|--------|-----
| `next`       | Integration branch. new features are merged directly here. Released to master once stable
| feature-name  | Feature branches. Feature development happens here. merged to `next`
| patch-name | Bugfix branches. merged directly to master
| master | Latest branch. Reflects current state of the published package

Again, this is my opinion. Take it as such.

**Comment by sahariko** [comment]
@omrilotan,
Is there a way we can automatically bump the package version on master, without requiring developers to remember to change it from `rc` to a clean semver?

@ealush,
Regarding the release flow you suggested - I think it involves too many steps, which are not as intuitive as the ones we employ right now.
It creates this "black magic" feeling, which estranges developers from the environment and how it works.

As to naming conventions - while I agree with having a `next`, I don't think it will serve us with most repos because it, as a concept, makes the version release process way slower and more meticuolous, which doesn't suit our agenda.

**Comment by ealush** [comment]
@sahariko 

image
A. Publishing a new version should NOT be simple, it should be explicit, and I would trust a bot any day to do a more decent job than I do. Ideally, developers shouldn't even be aware of the publication process. IMO, when the process matures, we should even graduate to automatic release cycles. But that's far ahead of us.

B. Again, the release-process shouldn't be fast. Why the rush to master? **Are we not done growth hacking?** Build a version, test it in your integration branch (next), consume it using `next` version for a while during development, and when you are done making changes, and feel like your feature is stable, merge to master and publish your new version.

**Comment by sahariko** [comment]
@ealush,
A. I completely agree with making this process as seamless as possible, but the flow you described doesn't make it any more explicit or transparent, on the contrary, it makes it obscure.
Also, I'm not a big fan not trusting developers with their code (a la `dangerouslySetInnerHTML`), especially when it's in-house.

B. Cumulative versioning is not a one-fix-suits-all IMO. While it might be good for more extensive libraries/modules (for example, `globals.js` or `fit`), I think we should preserve the power to develop fast with our webapps.

**Comment by michael5r** [comment]
@ealush The goal right now is to find a mix of publishing fast & publishing smart - eg. we very much want people to be able to run fast here (because we're still in the early stages of v4), but we also want safeguards in place to stop stupid mistakes from happening.

As such, we do want the publishing process to be fairly simple at this stage - we don't want to throw artificial hurdles in the way of the developers, simply to slow things down, nor can I ever envision a time where Fiverr moves to automatic release cycles (we're actually quite proud of the fact that we aren't doing those).

@omrilotan 's solution is a great fit for how we need to be working right now - it doesn't mean that we can't further discuss (or change) it later on, though, when we're past the initial gold rush of v4 development & our use of these technologies have matured.
